File: android/idsm_manager/app/src/main/java/com/idsm/manager/VsocEnvelope.kt
package com.idsm.manager

import org.json.JSONArray
import org.json.JSONObject
import java.security.MessageDigest

/**
 * VsocEnvelope -- 检测日志上报信封构造(设计文档第 6/9 章)。
 *
 * 探针 UDS NDJSON 行 -> alert_{host|eth|can} 信封 content 条目:
 * eventId(确定性派生, 幂等键)/eventType/severity/timestamp/
 * ecuCode/nodeType/ruleVersion/replay/raw(原始行透传)。
 *
 * 与 linux/idsm_managerd/vsoc_envelope.cpp 逐字段一致, 同一 mock 云
 * (tools/vsoc_mock/mock_vsoc.py 校验器)对两侧一视同仁。
 */
object VsocEnvelope {

    data class ParsedAlert(
        val eventId: Long,          /* IdsRM 外发事件 id(如 0x8025) */
        val severity: String,       /* LOW/MEDIUM/HIGH/CRITICAL */
        val timestampMs: Long,      /* 探针时间戳换算 */
        val idsMessageHex: String,  /* 完整 IDSM 消息 hex, eventId 派生源 */
    )

    /** HIDPS 事件类型名(0x8021-0x802A, 与 host_probe 检测器一一对应) */
    private val hostEventNames = mapOf(
        0x8021L to "DT_UNKNOWN_EXEC",
        0x8022L to "DT_PRIV_ESC",
        0x8023L to "DT_FORK_FLOOD",
        0x8024L to "DT_REVERSE_SHELL",
        0x8025L to "DT_FILE_MOD",
        0x8026L to "DT_NEW_SETUID",
        0x8027L to "DT_KMOD",
        0x8028L to "DT_ZOMBIE_STORM",
        0x8029L to "DT_RES_EXHAUST",
        0x802AL to "DT_ROOT_SHELL",
    )

    fun parseAlertLine(line: String): ParsedAlert? = try {
        val j = JSONObject(line)
        val idsMessage = j.optString("ids_message")
        ParsedAlert(
            eventId = j.getLong("event_id"),
            severity = j.optString("severity", "MEDIUM"),
            timestampMs = j.optLong("timestamp_s") * 1000 +
                    j.optLong("timestamp_ns") / 1_000_000,
            /* 无 ids_message 时退化为整行派生, 保证 eventId 仍可去重 */
            idsMessageHex = if (idsMessage.isEmpty()) sha256Hex(line)
                            else idsMessage,
        )
    } catch (e: Exception) {
        null
    }

    fun eventTypeName(nodeType: String, eventId: Long): String {
        if (nodeType == "HIDPS") {
            hostEventNames[eventId]?.let { return it }
        }
        return "EVT_%04X".format(eventId)
    }

    /** 确定性 eventId: {device_id}-{nodeType}-{sha256(ids_message_hex) 前 32 hex} */
    fun makeEventId(deviceId: String, nodeType: String, idsMessageHex: String): String =
        "$deviceId-$nodeType-${sha256Hex(idsMessageHex).take(32)}"

    /**
     * 批量构造 alert 信封; 全部行非法时返回 null(调用方丢批推进, 防毒丸卡死队列)。
     */
    fun buildAlertEnvelope(
        nodeType: String,
        ecuCode: String,
        ruleVersion: String,
        rawLines: List<String>,
    ): String? {
        val content = JSONArray()
        for (line in rawLines) {
            val a = parseAlertLine(line) ?: continue
            val raw = runCatching { JSONObject(line) }.getOrElse { JSONObject() }
            content.put(JSONObject().apply {
                put("eventId", makeEventId(DeviceIdentity.deviceId(), nodeType,
                                           a.idsMessageHex))
                put("eventType", eventTypeName(nodeType, a.eventId))
                put("severity", a.severity)
                put("timestamp", a.timestampMs)
                put("ecuCode", ecuCode)
                put("nodeType", nodeType)
                put("ruleVersion", ruleVersion)
                put("replay", false)
                put("raw", raw)
            })
        }
        if (content.length() == 0) return null
        val seg = DeviceIdentity.nodeTypeToTopicSeg(nodeType)
        if (seg == "unknown") return null
        return JSONObject().apply {
            put("msg_type", "alert_$seg")
            put("protocol_version", "1.0")
            put("timestamp", System.currentTimeMillis())
            put("manufacturer", DeviceIdentity.manufacturer)
            put("content", content)
        }.toString()
    }

    private fun sha256Hex(inBytes: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(inBytes.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }
}

File: linux/idsm_managerd/vsoc_envelope.cpp
#include "vsoc_envelope.h"

#include <nlohmann/json.hpp>

#include <openssl/evp.h>

#include <chrono>
#include <cstdio>
#include <map>

namespace idsm {

static std::string sha256Hex(const std::string& in) {
    uint8_t digest[EVP_MAX_MD_SIZE];
    unsigned int dlen = 0;
    EVP_Digest(in.data(), in.size(), digest, &dlen, EVP_sha256(), nullptr);
    std::string out;
    out.reserve(static_cast<size_t>(dlen) * 2);
    for (unsigned int i = 0; i < dlen; ++i) {
        char buf[3];
        std::snprintf(buf, sizeof(buf), "%02x", digest[i]);
        out += buf;
    }
    return out;
}

bool parseAlertLine(const std::string& line, ParsedAlert& out, std::string& err) {
    nlohmann::json j;
    try {
        j = nlohmann::json::parse(line);
    } catch (const std::exception& e) {
        err = std::string("json: ") + e.what();
        return false;
    }
    try {
        out.event_id = j.at("event_id").get<uint32_t>();
        out.severity = j.value("severity", std::string("MEDIUM"));
        const int64_t s = j.value("timestamp_s", int64_t{0});
        const int64_t ns = j.value("timestamp_ns", int64_t{0});
        out.timestamp_ms = s * 1000 + ns / 1000000;
        out.ids_message_hex = j.value("ids_message", std::string());
        if (out.ids_message_hex.empty()) {
            /* 无 ids_message 时退化为整行派生, 保证 eventId 仍可去重 */
            out.ids_message_hex = sha256Hex(line);
        }
    } catch (const std::exception& e) {
        err = std::string("field: ") + e.what();
        return false;
    }
    return true;
}

const char* hostEventTypeName(uint32_t event_id) {
    static const std::map<uint32_t, const char*> kNames = {
        {0x8021, "DT_UNKNOWN_EXEC"},
        {0x8022, "DT_PRIV_ESC"},
        {0x8023, "DT_FORK_FLOOD"},
        {0x8024, "DT_REVERSE_SHELL"},
        {0x8025, "DT_FILE_MOD"},
        {0x8026, "DT_NEW_SETUID"},
        {0x8027, "DT_KMOD"},
        {0x8028, "DT_ZOMBIE_STORM"},
        {0x8029, "DT_RES_EXHAUST"},
        {0x802A, "DT_ROOT_SHELL"},
    };
    const auto it = kNames.find(event_id);
    if (it != kNames.end()) return it->second;
    static thread_local char buf[16];
    std::snprintf(buf, sizeof(buf), "EVT_%04X", event_id);
    return buf;
}

std::string makeEventId(const std::string& device_id,
                        const std::string& node_type,
                        const std::string& ids_message_hex) {
    return device_id + "-" + node_type + "-" + sha256Hex(ids_message_hex).substr(0, 32);
}

const char* nodeTypeToTopicSeg(const std::string& node_type) {
    if (node_type == "HIDPS") return "host";
    if (node_type == "NIDPS") return "eth";
    if (node_type == "CIDS") return "can";
    return "unknown";
}

std::string buildEventUpEnvelope(const std::string& manufacturer,
                                 const std::vector<EventItem>& items) {
    const long long now_ms =
        std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::system_clock::now().time_since_epoch()).count();
    nlohmann::json content = nlohmann::json::array();
    for (const auto& it : items) {
        content.push_back({
            {"eventType", it.event_type},
            {"severity", it.severity},
            {"timestamp", it.timestamp_ms > 0 ? it.timestamp_ms : now_ms},
            {"detail", it.detail},
        });
    }
    const nlohmann::json env = {
        {"msg_type", "event_up"},
        {"protocol_version", "1.0"},
        {"timestamp", now_ms},
        {"manufacturer", manufacturer},
        {"content", std::move(content)},
    };
    return env.dump();
}

std::string buildAlertEnvelope(const std::string& manufacturer,
                               const std::string& device_id,
                               const std::string& node_type,
                               const std::string& ecu_code,
                               const std::string& rule_version,
                               const std::vector<std::string>& raw_lines,
                               std::string& err) {
    nlohmann::json content = nlohmann::json::array();
    size_t dropped = 0;
    for (const auto& line : raw_lines) {
        ParsedAlert a;
        std::string perr;
        if (!parseAlertLine(line, a, perr)) {
            ++dropped;
            continue;
        }
        nlohmann::json raw;
        try {
            raw = nlohmann::json::parse(line);
        } catch (...) {
            raw = nlohmann::json::object();
        }
        std::string event_type;
        if (node_type == "HIDPS") {
            event_type = hostEventTypeName(a.event_id);
        } else {
            char buf[16];
            std::snprintf(buf, sizeof(buf), "EVT_%04X", a.event_id);
            event_type = buf;
        }
        content.push_back({
            {"eventId", makeEventId(device_id, node_type, a.ids_message_hex)},
            {"eventType", event_type},
            {"severity", a.severity},
            {"timestamp", a.timestamp_ms},
            {"ecuCode", ecu_code},
            {"nodeType", node_type},
            {"ruleVersion", rule_version},
            {"replay", false},
            {"raw", std::move(raw)},
        });
    }
    if (content.empty()) {
        err = "no valid alert lines (" + std::to_string(dropped) + " dropped)";
        return {};
    }
    const nlohmann::json env = {
        {"msg_type", std::string("alert_") + nodeTypeToTopicSeg(node_type)},
        {"protocol_version", "1.0"},
        {"timestamp",
         std::chrono::duration_cast<std::chrono::milliseconds>(
             std::chrono::system_clock::now().time_since_epoch()).count()},
        {"manufacturer", manufacturer},
        {"content", std::move(content)},
    };
    return env.dump();
}

}  /* namespace idsm */
